### PR TITLE
r/aws_lambda_function Disable lambda from being triggered by setting reserved_concurrent_executions to 0

### LINKED ISSUE #3803: reserved_concurrent_executions for aws_lambda_function does not support 0 concurrency
### Actual Result
If `reserved_concurrent_executions = 0` for an `aws_lambda_function` in Terraform, reserve concurrency is disabled:

screen shot 2018-03-15 at 6 09 58 pm

### Expected Result
However, 0 is actually a valid value with a specific meaning, which you can see via the console:

screen shot 2018-03-15 at 6 10 07 pm

Why would you ever set reserved concurrency to 0? From the [AWS documentation](https://docs.aws.amazon.com/lambda/latest/dg/concurrent-executions.html):

_If you need a function to stop processing any invocations, you can choose to set the concurrency to 0 and throttle all incoming executions._

This is useful for long-running jobs (example: https://github.com/airbnb/binaryalert/issues/46)

### Requested Change
Provide a way to distinguish between _disabled_ reserve concurrency and _no_ reserve concurrency. Not sure the best approach here - maybe a negative value or an empty string disables it?

### Versions
Terraform 0.11.3 with AWS provider 1.11.0

### Affected Resource(s)
- aws_lambda_function

### Terraform Configuration Files
```hcl
resource "aws_lambda_function" "function" {
  function_name = "${var.function_name}"
  runtime       = "python2.7"
  role          = "${aws_iam_role.role.arn}"
  handler       = "${var.handler}"
  s3_bucket     = "${var.source_bucket}"
  s3_key        = "${var.source_object_key}"

  // Should temporarily throttle all invocations
  reserved_concurrent_executions = 0
}
```

### PR DESCRIPTION
**Pull request by Puneeth-n**
closes #3803 

```
AWS_REGION=us-east-1 make testacc TESTARGS='-run=TestAccAWSLambdaFunction_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccAWSLambdaFunction_ -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSLambdaFunction_importLocalFile
--- PASS: TestAccAWSLambdaFunction_importLocalFile (42.48s)
=== RUN   TestAccAWSLambdaFunction_importLocalFile_VPC
--- PASS: TestAccAWSLambdaFunction_importLocalFile_VPC (43.17s)
=== RUN   TestAccAWSLambdaFunction_importS3
--- PASS: TestAccAWSLambdaFunction_importS3 (44.41s)
=== RUN   TestAccAWSLambdaFunction_basic
--- PASS: TestAccAWSLambdaFunction_basic (37.53s)
=== RUN   TestAccAWSLambdaFunction_concurrency
--- PASS: TestAccAWSLambdaFunction_concurrency (82.14s)
=== RUN   TestAccAWSLambdaFunction_concurrencyCycle
--- PASS: TestAccAWSLambdaFunction_concurrencyCycle (104.85s)
=== RUN   TestAccAWSLambdaFunction_updateRuntime
--- PASS: TestAccAWSLambdaFunction_updateRuntime (59.85s)
=== RUN   TestAccAWSLambdaFunction_expectFilenameAndS3Attributes
--- PASS: TestAccAWSLambdaFunction_expectFilenameAndS3Attributes (26.76s)
=== RUN   TestAccAWSLambdaFunction_envVariables
--- PASS: TestAccAWSLambdaFunction_envVariables (101.95s)
=== RUN   TestAccAWSLambdaFunction_encryptedEnvVariables
--- PASS: TestAccAWSLambdaFunction_encryptedEnvVariables (79.26s)
=== RUN   TestAccAWSLambdaFunction_versioned
--- PASS: TestAccAWSLambdaFunction_versioned (53.19s)
=== RUN   TestAccAWSLambdaFunction_versionedUpdate
--- PASS: TestAccAWSLambdaFunction_versionedUpdate (59.37s)
=== RUN   TestAccAWSLambdaFunction_DeadLetterConfig
--- PASS: TestAccAWSLambdaFunction_DeadLetterConfig (38.38s)
=== RUN   TestAccAWSLambdaFunction_DeadLetterConfigUpdated
--- PASS: TestAccAWSLambdaFunction_DeadLetterConfigUpdated (59.80s)
=== RUN   TestAccAWSLambdaFunction_nilDeadLetterConfig
--- PASS: TestAccAWSLambdaFunction_nilDeadLetterConfig (26.67s)
=== RUN   TestAccAWSLambdaFunction_tracingConfig
--- PASS: TestAccAWSLambdaFunction_tracingConfig (59.21s)
=== RUN   TestAccAWSLambdaFunction_VPC
--- PASS: TestAccAWSLambdaFunction_VPC (40.62s)
=== RUN   TestAccAWSLambdaFunction_VPCUpdate
--- PASS: TestAccAWSLambdaFunction_VPCUpdate (71.15s)
=== RUN   TestAccAWSLambdaFunction_VPC_withInvocation
--- PASS: TestAccAWSLambdaFunction_VPC_withInvocation (100.99s)
=== RUN   TestAccAWSLambdaFunction_s3
--- PASS: TestAccAWSLambdaFunction_s3 (42.01s)
=== RUN   TestAccAWSLambdaFunction_localUpdate
--- FAIL: TestAccAWSLambdaFunction_localUpdate (38.58s)
	testing.go:518: Step 1 error: Check failed: Check 5/5 error: Expected time attribute data.template_file.last_modified.rendered with value 2018-03-19T13:45:03.108+0000 was not before 2018-03-19T13:45:05Z
=== RUN   TestAccAWSLambdaFunction_localUpdate_nameOnly
--- PASS: TestAccAWSLambdaFunction_localUpdate_nameOnly (45.44s)
=== RUN   TestAccAWSLambdaFunction_s3Update_basic
--- PASS: TestAccAWSLambdaFunction_s3Update_basic (95.66s)
=== RUN   TestAccAWSLambdaFunction_s3Update_unversioned
--- PASS: TestAccAWSLambdaFunction_s3Update_unversioned (70.01s)
=== RUN   TestAccAWSLambdaFunction_runtimeValidation_noRuntime
--- PASS: TestAccAWSLambdaFunction_runtimeValidation_noRuntime (1.67s)
=== RUN   TestAccAWSLambdaFunction_runtimeValidation_nodeJs43
--- PASS: TestAccAWSLambdaFunction_runtimeValidation_nodeJs43 (37.11s)
=== RUN   TestAccAWSLambdaFunction_runtimeValidation_python27
--- PASS: TestAccAWSLambdaFunction_runtimeValidation_python27 (36.68s)
=== RUN   TestAccAWSLambdaFunction_runtimeValidation_java8
--- PASS: TestAccAWSLambdaFunction_runtimeValidation_java8 (39.05s)
=== RUN   TestAccAWSLambdaFunction_tags
--- PASS: TestAccAWSLambdaFunction_tags (82.64s)
=== RUN   TestAccAWSLambdaFunction_runtimeValidation_python36
--- PASS: TestAccAWSLambdaFunction_runtimeValidation_python36 (37.54s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	1658.207s
GNUmakefile:18: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

### CODE REVIEW

**Comment by austinbyers** [comment]
Thanks for the change @Puneeth-n! Just keep in mind that people might be relying on 0 as the default value right now, so this is a **breaking change**.

In other words, if someone has a Lambda function with 0 concurrency in the current provider version, their Lambda will run without any limit. But after this change, their existing config (with a 0 value) will stop all invocations of their function

I imagine this would be a fairly rare example, but this just needs to be clearly demarcated in the release notes

File: aws/resource_aws_lambda_function.go [hunk 1]
```diff
@@ -820,6 +821,8 @@ func resourceAwsLambdaFunctionUpdate(d *schema.ResourceData, meta interface{}) e
 			if err != nil {
 				return fmt.Errorf("Error setting concurrency for Lambda %s: %s", d.Id(), err)
 			}
+		} else {
+			return fmt.Errorf("Error processing concurrency for Lambda %s. reserved_concurrent_executions should be >=0 for limiting concurrency, reserved_concurrent_executions = 0 for disabling this lambda from being triggered and reserved_concurrent_executions = -1 for removing concurrency constraint. Supplied value: %d", d.Id(), nc.(int))
```
Comment (bflad) [inline aws/resource_aws_lambda_function.go]: This can be removed with `ValidateFunc: validation.IntAtLeast(-1),` on the attribute 👍

Comment (Puneeth-n) [reply aws/resource_aws_lambda_function.go]: @bflad Thanks for the review :) Will make the necessary changes.

**Comment by Puneeth-n** [comment]
@bflad used the validate function from validators. I noticed that an unrelated test is failing on master. 

```
AWS_REGION=us-east-1 make testacc TESTARGS='-run=TestAccAWSLambdaFunction_localUpdate'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccAWSLambdaFunction_localUpdate -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSLambdaFunction_localUpdate
--- FAIL: TestAccAWSLambdaFunction_localUpdate (39.34s)
	testing.go:513: Step 1 error: Check failed: Check 5/5 error: Expected time attribute data.template_file.last_modified.rendered with value 2018-03-19T13:20:28.580+0000 was not before 2018-03-19T13:20:30Z
=== RUN   TestAccAWSLambdaFunction_localUpdate_nameOnly
--- PASS: TestAccAWSLambdaFunction_localUpdate_nameOnly (44.93s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	84.296s
GNUmakefile:18: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

**Comment by Puneeth-n** [comment]
@radeksimko @Ninir  Cab you please review it? :)

**Comment by bflad** [comment]
@Puneeth-n sorry this got lost in the shuffle 😅 Could you rebase this and we'll get it in? Thanks so much.

**Comment by Puneeth-n** [comment]
@bflad Finally :D Sure. Will fix it!

**Comment by bflad** [comment]
We're starting to see some duplicate issues/pull requests for this functionality: #5288 

I'm still worried that since this could potentially cause production downtime (due to the nature of the change) that we might need to relegate this merge to only happen during the next major version upgrade of the provider, likely in the coming weeks before Terraform 0.12. For now, I'm going to tag this with the `breaking-change` label and `2.0.0` milestone just to ensure its on our radar for that timeframe and denote its severity.

**Comment by zioproto** [comment]
Is this ever going to be merged in Terraform 0.11.x ? thanks

**Pull request merged**